### PR TITLE
Switch to typescript

### PR DESCRIPTION
**Pull request by abumalick**
Would you accept a PR to switch to typescript?

I can make it more readable inchallah. I was a bit aggressive on this one:
- I formatted most files using Prettier
- I switched to pika to build the package
- I removed the build file from the repository

I can remove some of these and make it more readable (commit after renaming js files to ts so that you can check actual changes).

I wanted to ask you before working on this actually.

### CODE REVIEW

**Comment by abumalick** [comment]
@z3bi السلام عليكم
What do you think about this?

**Comment by z3bi** [comment]
@abumalick this is awesome! I'm 100% onboard with turning this into typescript. 

I'm not familiar with pika, what are the advantages of that compared to other package managers?

I think it should be ok to remove the built file from the repo, but we would want to update the ci pipeline to generate an artifact with releases. We will probably also want to keep the import the same if possible to avoid breaking changes.

**Comment by abumalick** [comment]
Happy that you are onboard!

You can read on pika here for example: https://www.pika.dev/blog/introducing-pika-pack
Basically it simplifies the configuration for creating packages. Also @FredKSchott that is maintaining is working a lot on having packages supporting new browser features, so they will keep optimizing it without needing to us to dig into the building process. We can just use it and not think too much about it. It is using babel and rollup but is simplifying the configuration for packages maintainers.

It is noted for the built file, I will check that in the end.

I will hard push more organized commits to show you the actual changes to the files. I am trying to not have any breaking changes.

File: .github/workflows/node.js.yml [hunk 1]
```diff
@@ -83,14 +82,13 @@ jobs:
     # Steps represent a sequence of tasks that will be executed as part of the job
     steps:
       - uses: actions/checkout@v2
-        with:
-          token: ${{ secrets.GIT_ACCESS_TOKEN }}
```
Comment (abumalick) [inline .github/workflows/node.js.yml:87]: I believe this was not used, I don't really understand why it is there actually

File: .github/workflows/node.js.yml [hunk 2]
```diff
@@ -17,20 +17,17 @@ jobs:
 
     strategy:
       matrix:
-        node-version: [10.x, 12.x, 14.x, 15.x]
```
Comment (abumalick) [inline .github/workflows/node.js.yml:20]: I am dropping version 10 and 15 which is not supported anymore https://nodejs.org/en/about/releases/

**Comment by abumalick** [comment]
@z3bi You can check https://github.com/batoulapps/adhan-js/pull/93/commits/788bd2d35f7d941087f31d8cb47650e9467734a7 to see the exact changes to the files that I pushed to make it work with typescript.

I dropped pika pack in the end, in favor of https://github.com/developit/microbundle (pika didn't have a new release for 2 years). It does basically the same, it handles most of the configuration by itself.

> I think it should be ok to remove the built file from the repo, but we would want to update the ci pipeline to generate an artifact with releases. We will probably also want to keep the import the same if possible to avoid breaking changes.

I added the build to the release job. It should work, I didn't test it yet through, the pipeline passes in my github but I didn't increase the version so no release for now. https://github.com/abumalick/adhan-js/actions/runs/2089942331

So the only change to the final package is that the built files are inside the dist folder, we have now multiple build files (commonJs, umd, etc...)

```json
"exports": {
    "require": "./dist/Adhan.cjs",
    "default": "./dist/Adhan.modern.js"
  },
  "main": "./dist/Adhan.cjs",
  "umd:main": "dist/Adhan.umd.js",
  "module": "./dist/Adhan.module.js",
  "unpkg": "./dist/Adhan.umd.js",
  "types": "./dist/Adhan.d.ts",
```

I believe it should not break anyone's usage of the package, but I would at least bump a minor version as it has important changes and might break in some extreme cases. What do you think? Should I bump to 4.4.0 or 5.0.0?

I will try to test the release assets when you tell me what is the version that I should use inchallah

File: .github/workflows/node.js.yml (hunk 2 above)
Comment (abumalick) [reply .github/workflows/node.js.yml:20]: it allows us to upgrade to npm 8

**Comment by abumalick** [comment]
@z3bi Do you have time to check my comments please ?

**Comment by sgtsquiggs** [comment]
Hi @abumalick, I've already got a typescript rewrite in progress. When it is ready I'll post it and we can compare/contrast.

**Comment by abumalick** [comment]
This PR is not a rewrite, @sgtsquiggs , it is more a transition to typescript. I only added the types that are necessary and didn't change the code structure.
There is some refactoring that could be done for a prettier typescript version, like use enums instead of objects.

You can check the typescript adaptation in this commit: https://github.com/batoulapps/adhan-js/pull/93/commits/6f17cae1ddabdbbd5e7b1fbc1f892ed12ab3357c

**Comment by abumalick** [comment]
@z3bi Any follow up?

**Comment by sgtsquiggs** [comment]
This would be easier to review if the js -> ts conversion was registered as a diff instead of a delete/create - possibly the eslint/prettier rules caused too many changes for git to automatically associate the two files as related. Could you submit prettier/eslint change as a separate MR that we could merge before this ts MR?

**Comment by sgtsquiggs** [comment]
Also can you explain the need to rename the config file extensions `js` -> `cjs` ? This seems uncommon to me.

**Comment by sgtsquiggs** [comment]
For packaging - instead of webpack/pika/microbundle I would prefer to use rollup. I was in the process of updating our bundling:

`rollup.config.js`
```js
import babel from "@rollup/plugin-babel";
import resolve from "@rollup/plugin-node-resolve";
import { terser } from "rollup-plugin-terser";

const extensions = [".js", ".ts"];

export default {
  input: "src/Adhan.js",
  output: [
    {
      file: "lib/bundles/bundle.esm.js",
      format: "esm",
      sourcemap: true,
    },
    {
      file: "lib/bundles/bundle.esm.min.js",
      format: "esm",
      plugins: [terser()],
      sourcemap: true,
    },
    {
      file: "lib/bundles/bundle.umd.js",
      format: "umd",
      name: "myLibrary",
      sourcemap: true,
    },
    {
      file: "lib/bundles/bundle.umd.min.js",
      format: "umd",
      name: "myLibrary",
      plugins: [terser()],
      sourcemap: true,
    },
  ],
  plugins: [
    resolve({ extensions }),
    babel({
      babelHelpers: "bundled",
      exclude: "./node_modules/**",
      extensions,
      include: ["src/**/*.js"],
    }),
  ],
};
```

With these `package.json` scripts:
```js
    "clean": "rimraf lib",
    "build:esm": "cross-env BABEL_ENV=esmUnbundled babel src --extensions '.js' --out-dir 'lib/esm' --source-maps",
    "build:cjs": "cross-env BABEL_ENV=cjs babel src --extensions '.js' --out-dir 'lib/cjs' --source-maps",
    "build:bundles": "cross-env BABEL_ENV=esmBundled rollup -c",
    "build": "npm-run-all -l clean -p build:esm build:cjs build:bundles",
```

**Comment by abumalick** [comment]
@sgtsquiggs @z3bi Can I drop the Math.trunc polyfill that is in Astronomical file?
image

**Comment by abumalick** [comment]
I believe this PR is ready for review @z3bi @sgtsquiggs 🚀

**Comment by sgtsquiggs** [comment]
Can you `npm remove -D @types/estree @types/jest` I believe these are unnecessary

**Comment by sgtsquiggs** [comment]
> @sgtsquiggs @z3bi Can I drop the Math.trunc polyfill that is in Astronomical file?
> image

100%

File: jest.config.js [hunk 3]
```diff
@@ -1,7 +1,7 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
```
Comment (sgtsquiggs) [inline jest.config.js:1]: is this necessary?

**Review by sgtsquiggs — COMMENTED** [review]
.eslint.js can be replaced with:
```javascript
module.exports = {
  parser: '@typescript-eslint/parser',

  parserOptions: {
    ecmaVersion: 2018,
    project: ['./tsconfig.json'],
    sourceType: 'module',
  },

  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:jest/recommended',
    'plugin:jest/style',
    'plugin:prettier/recommended',
  ],

  overrides: [
    {
      files: ['*.js'],
      parser: null,
      parserOptions: { project: null },
      rules: {
        '@typescript-eslint/ban-ts-comment': 'off',
        '@typescript-eslint/no-unnecessary-condition': 'off',
      },
    },
  ],

  env: {
    node: true,
    browser: true,
    'jest/globals': true,
  },

  rules: {
    '@typescript-eslint/ban-ts-comment': 'warn',
    '@typescript-eslint/no-unnecessary-condition': 'warn',
    complexity: ['warn', 10],
    'max-lines': ['warn', 300],
    'max-params': ['warn', 5],
    eqeqeq: ['error', 'smart'],
    'no-var': 'error',
    'prefer-const': 'error',
    'object-shorthand': 'error',
  },

  plugins: ['@typescript-eslint', 'jest'],
};
```

and then you can drop the second tsconfig

File: jest.config.js (hunk 3 above)
Comment (abumalick) [reply jest.config.js:1]: it is not required, it is from official documentation and give you some completion.
image

Do you want to remove it?

Comment (sgtsquiggs) [reply jest.config.js:1]: Oh nice 👍

**Review by sgtsquiggs — CHANGES_REQUESTED** [review]
Last thing - can you  `npm remove -D @types/estree @types/jest` I believe these are unnecessary

**Comment by sgtsquiggs** [comment]
> I believe `@types/jest` is necessary to give you typescript support for jest utilities in test files.

👍 

> For `@types/estree`, I am getting an error when it is not there, I can try again inchallah but it is the only way I found to fix that error.

I had ran into the same estree error as you - I was able to get it to stop appearing without adding `@types/estree` but I forgot what exactly I did. Perhaps we just merge this and fix later.

**Review by sgtsquiggs — APPROVED** [review]

**Comment by abumalick** [comment]
Thank you very much @sgtsquiggs 👍🏻

**Pull request merged**